core/src/token_info/cached.rs

```diff
@@ -0,0 +1,159 @@
+use super::{TokenBaseInfo, TokenId, TokenInfoFetching};
+
+use anyhow::Result;
+use futures::future::{BoxFuture, FutureExt};
+use std::collections::HashMap;
+use std::sync::{Arc, RwLock};
+
+/**
+ * Implementation of TokenInfoFetching that stores previously fetched information in an in-memory cache for fast retrieval.
+ * TokenIds will always be fetched from the inner layer, as new tokens could be added at any time.
+ */
+struct TokenInfoCache {
+    cache: RwLock<HashMap<TokenId, TokenBaseInfo>>,
+    inner: Arc<dyn TokenInfoFetching>,
+}
+
+impl TokenInfoCache {
+    #[allow(dead_code)]
+    pub fn new(inner: Arc<dyn TokenInfoFetching>) -> Self {
+        Self {
+            cache: RwLock::new(HashMap::new()),
+            inner,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn new_with_existing_cache(
```

Comment:
ubernit: Typically, Rust functions that create a type with options, is just named `with_*` (like `Vec::new()` and `Vec::with_capacity()`). While its not a big deal at all, I think it would be better to follow the convention here.